### ISSUE TITLE
Error occurred while "Rolling back fixes...". How do I recover pre-commit stash?

### ISSUE THREAD

**Issue by czuba** [pre-commit/pre-commit-hooks#1035]
## TL;DR
Pre-commit **_ate my uncommitted changes_** within its own stash!
- How do I recover the state of my repo prior to the [failed] commit attempt?
  - i.e. I need to apply the stash that pre-commit created, but it doesn't appear in my `git stash list`
- How can this be prevented from happening again?
  - Is recovery from similar circumstance already addressed in an FAQ somewhere?  (...none that I can find)

## Background:
Encountered an error while attempting to commit a minor file in a repo that had many other changes. After erroring out, **_all my uncommitted changes are missing!_**

Pre-commit status:
- lots of existing changes in repo
- made minor revision [to .gitignore] that I wanted to separately commit
- executed `git add .gitignore` & `git commit`
- pre-commit hooks ran with the following output:

``` bash
(condaEnv) myRepo [dev *]🌮 git add .gitignore
(condaEnv) myRepo [dev *+]🌮 git commit
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to <userHome>/.cache/pre-commit/patch1712082453-71444.
jupytext.............................................(no files to check)Skipped
[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
An unexpected error has occurred: CalledProcessError: command: ('/usr/local/git/libexec/git-core/git', '-c', 'core.autocrlf=false', 'apply', '--whitespace=nowarn', '<userHome>/.cache/pre-commit/patch1712082453-71444')
return code: 1
stdout: (none)
stderr:
    <userHome>/miniforge3/envs/<condaEnv>/lib/python3.11/site-packages/nbformat/__init__.py:93: MissingIDFieldWarning: Code cell is missing an id field, this will become a hard error in future nbformat versions. You may want to use `normalize()` on your notebooks before validations (available since nbformat 5.1.4). Previous versions of nbformat are fixing this issue transparently, and will stop doing so in the future.
      validate(nb)
    error: patch failed: <dirtyFile_notInCommit>.ipynb:66
    error: <dirtyFile_notInCommit>.ipynb: patch does not apply
Check the log at <userHome>/.cache/pre-commit/pre-commit.log
<userHome>/miniforge3/envs/<condaEnv>/lib/python3.11/site-packages/nbformat/__init__.py:93: MissingIDFieldWarning: Code cell is missing an id field, this will become a hard error in future nbformat versions. You may want to use `normalize()` on your notebooks before validations (available since nbformat 5.1.4). Previous versions of nbformat are fixing this issue transparently, and will stop doing so in the future.
  validate(nb)
<userHome>/miniforge3/envs/<condaEnv>/lib/python3.11/site-packages/nbformat/__init__.py:93: MissingIDFieldWarning: Code cell is missing an id field, this will become a hard error in future nbformat versions. You may want to use `normalize()` on your notebooks before validations (available since nbformat 5.1.4). Previous versions of nbformat are fixing this issue transparently, and will stop doing so in the future.
  validate(nb)
(condaEnv) myRepo [dev *+]🌮 
```

## My uncommitted changes were not reverted from the pre-commit stash!

`git status` after the error only shows:

``` bash
(<condaEnv>) myRepo [dev *+]🌮 git status
On branch dev
Your branch is up to date with 'origin/dev'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   .gitignore

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   .gitattributes

```
_...and nothing appears in the normal `git stash list`_

---
## Pre-commit config
My pre-commit config is [extremely] simple, just one hook (and even the default sample hooks are inactive).
- hook source:   https://jupytext.readthedocs.io/en/latest/using-pre-commit.html
- NOTE:  the hook itself doesn't appear to have had anything to do with the error state
  - ...the hook _would have_ triggered on the same `<dirtyFile_notInCommit>.ipynb`, but that file wasn't in the commit, so it didn't do anything **"`Skipped`"**

pre-commit-config.yaml:
``` pre-commit-config.yaml
repos:
#     # basic/default set of hooks (from pre-commit maintainers)
# -   repo: https://github.com/pre-commit/pre-commit-hooks
#     rev: v3.6.0
#     hooks:  
#     -   id: check-yaml
#     -   id: end-of-file-fixer
#     -   id: trailing-whitespace

    # Apply Jupytext on commit:
-   repo: https://github.com/mwouts/jupytext
    rev: v1.16.1  # CURRENT_TAG/COMMIT_HASH
    hooks:
    - id: jupytext
      args: [--sync]
```

## Suspicions
I think the source of the problem is
1. When stashing all local changes, **_a [misbehaving] filter that I had commented out_** in .gitattributes **_was inadvertently reactivated_**
   - `nbformat` --which throws the error-- is called by a git 'clean' filter that removes jupyter notebook execution count & output from the notebook file "`<dirtyFile_notInCommit>.ipynb`"
   -  ...its slow & error prone to notebook formatting issues that I don't care about, thus had been commented out in my local repo
2. It doesn't appear that any issue occurred with the pre-commit hook itself; _only with consequences of pre-commit stashing all local changes_ , irrespective of their staging status in the commit.
3. One of the stashed files --_which had nothing to do with the commit_-- that had been altered by the errantly reactivated .git filter now conflicted with the version within the pre-commit error state of the repo _(!!!)_
4. pre-commit returned its own dirty state of the repo without any trace of my stashed changes

**Comment by asottile** [comment]
see the first line of the output!

**Comment by asottile** [comment]
you've also reported this to the wrong project (pre-commit-hooks is a bunch of tools, pre-commit is the framework), if you had searched in the correct one there are duplicates including one specifically for the bug in git with smudge filters

**Comment by czuba** [comment]
Whoa, maybe dial it back a bit, @asottile. 

1. The pre-commit framework _did the stashing_, did it not? (Honestly, did it not?...feel free to say "no", but don't jump down users' throat about it) If a framework is going to hide away a bunch of user content that isn't relevant to the task at hand (in either a direct sense, or a user's experience sense), then that framework has a responsibility to not leave the environment in a state where said content is squirreled away in a separate location.
2. The first line of the output is the line that produced the error. _This is not the solution_, but it _is_ a direction toward a user-generated fix to the problem. (...indeed, what I had already used to solve the problem myself).
3. I made an effort to clearly state that it did not seem that _the hook itself_ was the culprit, but rather the interaction of pre-commit exposing code that had been inaccessible prior to stashing _all_ uncommitted changes —regardless of relevance to the commit— and said code leading to a failure mode that had not been accounted for/anticipated by the pre-commit framework.
